matrix-commander: init at unstable-2021-04-18

seb314 opened the pull request:
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`) (tested the python script in bin/matrix-commander, sending messages works)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

seb314 committed:
matrix-commander: init

r-rmcgibbo commented:
Result of `nixpkgs-review pr 119895` at c26e0eda run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>matrix-commander</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:23:3:
```
   |
23 |   installPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:32:5:
```
   |
32 |     license = lib.licenses.gpl3;
   |     ^
```
</li>
  </ul>
</details>

Mindavi commented:
LGTM. Builds on my machine and runs (I haven't actually tried connecting to a matrix server though).

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:
```diff
@@ -0,0 +1,36 @@
+{ pkgs, fetchFromGitHub, lib }:
+
+pkgs.stdenv.mkDerivation {
+  name = "matrix-commander";
```
```suggestion
  pname = "matrix-commander";
  version = "unstable-XXXX-XX-XX"
```

replace XXXX-XX-XX with the date of the commit.

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:
```diff
@@ -0,0 +1,36 @@
+    sha256 = "1cny7z793ln4xgf7cbgfd0jd4hi8mp8qr91siqjhy0nmyd9lmszv";
+  };
+
+  propagatedBuildInputs = [
+    pkgs.cacert ] ++
+    (with pkgs.python38Packages; [
+      matrix-nio
+      magic
+      markdown
+      pillow
+      urllib3
+    ]);
```
```suggestion
  propagatedBuildInputs = [ cacert ]
    ++ (with python3Packages; [
      matrix-nio
      magic
      markdown
      pillow
      urllib3
    ]);
```

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:
```diff
@@ -0,0 +1,36 @@
+{ pkgs, fetchFromGitHub, lib }:
```
```suggestion
{ lib, fetchFromGitHub, cacert, python3Packages }:
```

seb314 replied:
I also added stdenv in order to get mkDerivation -- or is there a better way to do that?

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp $src/matrix-commander.py $out/bin/matrix-commander
```
```suggestion
    mv $src/matrix-commander.py $out/bin/matrix-commander
```

seb314 replied:
with mv I get an error because src seems to be read-only...

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:
```diff
@@ -0,0 +1,36 @@
+    chmod +x $out/bin/matrix-commander
+  '';
+
+  meta = {
+    description = "Simple but convenient CLI-based Matrix client app for sending and receiving";
+    homepage = "https://github.com/8go/matrix-commander";
+    license = lib.licenses.gpl3;
+    platforms = lib.platforms.linux;
+    maintainers = [ lib.maintainers.seb314 ];
```
```suggestion
  meta = with lib; {
    description = "Simple but convenient CLI-based Matrix client app for sending and receiving";
    homepage = "https://github.com/8go/matrix-commander";
    license = licenses.gpl3;
    platforms = platforms.linux;
    maintainers = [ maintainers.seb314 ];
```

gpl3Only or gpl3Plus?

seb314 replied:
[asked](https://github.com/8go/matrix-commander/issues/29)

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -23946,6 +23946,8 @@ in
       canonicaljson;
   };
 
+  matrix-commander = callPackage ../applications/networking/instant-messengers/matrix-commander {};
```
```suggestion
  matrix-commander = callPackage ../applications/networking/instant-messengers/matrix-commander { };
```

seb314 committed:
matrix-commander: cleanups & potential dep fixes

apply most SuperSandro2000's review comments;

used python3.withPackages -- this might be the correct
way to get runtime python dependencies to work even
when matrix-commander is a dependency of another packages
(as opposed to testing in nix-shell). I'm not sure though.

aiofiles seemed to be a missing runtime dependency when called
from another (nix-packaged, jvm-based) app -- there was an
import error without it.

cacerts might require explicit pinning, as described here:
https://gist.github.com/CMCDragonkai/1ae4f4b5edeb021ca7bb1d271caca999
(relevant snippet: wrapProgram $out/bin/program \
  --set SSL_CERT_FILE "${cacert}/etc/ssl/certs/ca-bundle.crt"
  )

seb314 commented:
thanks for the review! In addition to applying most of your suggestions (see comments), I did some changes that I hope will be more correct regarding runtime dependencies (it seems that while the original commit worked fine from nix-shell, it resulted in an import error when I tried using matrix commander as a dependency from another nix-packaged app (jvm-based). The new version works locally from nix shell (as before) and avoids the import error when called from the other program. Hovewer, when called from my other app, actuall sending still doesn't work -- but that might be a problem on the other program/package's side...

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:
```diff
@@ -0,0 +1,38 @@
+  meta = with lib; {
+    description = "Simple but convenient CLI-based Matrix client app for sending and receiving";
+    homepage = "https://github.com/8go/matrix-commander";
+    license = licenses.gpl3;
```
gpl3Plus or gpl3Only?

seb314 replied:
I didn't find any mentions about that in the matrix-commander repo, so I [asked on github](https://github.com/8go/matrix-commander/issues/29), but there's no answer yet. Should wait, infer gpl3Only, or keep gpl3 for now?

Mindavi replied:
I think keeping it gpl3 for now and adding a link to the issue is the best way to handle this. Then it's easy to change in the future if / when an answer comes.

seb314 replied:
ok, added a note in [1331ee3](https://github.com/NixOS/nixpkgs/pull/119895/commits/1331ee3fae7930840714173a3bdec79a528ec54b)

seb314 committed:
add runHook {pre,post}Install

(suggested by r-rmcgibbo)

seb314 committed:
add link to github issue re gpl3{Only,Plus}

SuperSandro2000 commented on pkgs/applications/networking/instant-messengers/matrix-commander/default.nix:
```diff
@@ -0,0 +1,43 @@
+  meta = with lib; {
+    description = "Simple but convenient CLI-based Matrix client app for sending and receiving";
+    homepage = "https://github.com/8go/matrix-commander";
+    # currently unknown if gpl3Only or gpl3Plus: https://github.com/8go/matrix-commander/issues/29
+    license = licenses.gpl3;
```
```suggestion
    license = licenses.gpl3Only;
```

It is gpl3Only because there is no mention of later anywhere.

SuperSandro2000 committed:
Update pkgs/applications/networking/instant-messengers/matrix-commander/default.nix
